salvo/src/lib/benchmark/base_benchmark.py

```diff
@@ -78,20 +91,54 @@ def run_image(self, image_name : str, **kwargs) -> bytearray:
           here https://docker-py.readthedocs.io/en/stable/containers.html#docker.models.containers.ContainerCollection.run
 
     Returns:
-        A bytearray containing the output produced from executing the specified container
+        A bytearray containing the output produced from executing the specified
+        container
     """
-    pass
+    return self._docker_image.run_image(image_name, **kwargs)
 
   def pull_images(self) -> List[str]:
     """Retrieve all images necessary for the benchmark.
 
-    Retrieve the NightHawk and Envoy images defined in the control object. 
+    Retrieve the NightHawk and Envoy images defined in the control
+    object.
     """
-    pass
+    retrieved_images = []
+    images = self.get_images()
+
+    for image in [
+        images.nighthawk_benchmark_image,
+        images.nighthawk_binary_image,
+        images.envoy_image
+    ]:
+      # If the image name is not defined, we will have an empty string.
+      # For unit testing we'll keep this behavior. For true usage, we
+      # should raise an exception when the benchmark class performs its
+      # validation
+      if image:
+        i = self._docker_image.pull_image(image)
+        log.debug(f"Retrieved image: {i} for {image}")
+        if i is None:
+          return []
+        retrieved_images.append(i)
+
+    return retrieved_images
 
   def set_environment_vars(self) -> None:
     """Build the environment variable map used to launch an image.
 
-    Set the Envoy IP test versions and any other environment variables needed by the test
+    Set the Envoy IP test versions and any other environment variables needed
```

Comment:
(pre-existing) Does it make sense for this method to be public? Are there any run modes in which we wouldn't want to setup the environment?

Reply:
Two of the benchmark modes need to set `ENVOY_IP_TEST_VERSIONS`  and others would use this method to set other specified environment variables inherent to that benchmark (eg, `HEAPPROFILE`, `HEAPCHECK`, `ENVOY_PATH`).

Do you feel it should be private?

Reply:
That does sound like we have two different run modes. The reason I feel something is wrong with this method is that it is exposed on the API, but takes no arguments or parameters. So it feels like we only get a choice of calling it or not calling it. Looks like all the run modes you described will be calling it.

Since some of the benchmark modes actually need it to set other variables, how do they indicate to this method that other variables should be set? (since this method takes no arguments). Is it by setting `self._control.environment`? Where or what sets the `self._control.environment` attribute? It doesn't seem to be initialized anywhere.

If they do control the variables to be set by setting the `self._control.environment`, I would assume there will be a method that does that. Once we have such a method, maybe that can be combined with this one and there will no longer be a need for this as is? Not sure if I am explaining my confusion correctly, feel free to reach out on slack if I am not.

Reply:
From the doc [here](https://github.com/envoyproxy/nighthawk/blob/master/benchmarks/README.md), there are three execution modes.

This method operates on the values set in `self._control.environment` by the user.  

```
$ cat local_fully_dockerized_sources_only.json
{
  "remote": false,
  "dockerizedBenchmark": true,
  "images": {
    "reuseNhImages": true,
    "nighthawkBenchmarkImage": "envoyproxy/nighthawk-benchmark-dev:latest",
    "nighthawkBinaryImage": "envoyproxy/nighthawk-dev:latest"
  },
  "source": [{
    "identity": "SRCID_ENVOY",
    "source_url": "https://github.com/envoyproxy/envoy.git",
    "commit_hash": "d9af0e985a2fef2c921641b12f4786a658ad0923"
  }],
  "environment": {
    "testVersion": "IPV_V4ONLY",
    "envoyPath": "envoy",
    "outputDir": "/home/ubuntu/nighthawk_output",
    "testDir": "/home/ubuntu/nighthawk_tests"
  }
}
```

We look for the required variables that we must set, and any other free-form variables and values are specified in a map.   We call this method just before running the container, and I anticipate that each mode will be calling this method. 

I can ping you on slack a bit later to discuss.

Reply:
Thank you for the added details, there is something I don't understand here and talking on Slack may be the best. It is still unclear to me who and how sets the `self._control.environment`. Let's chat tomorrow on Slack, I am sure you will be able to help me identify what I misunderstood.